README.md
# check-aws-reservation-coverage

check aws reservation coverage

## Usage

```
Usage of build/linux_amd64/check-aws-reservation-coverage (0.1):
  -c float
    	set critical threshold (default 25)
  -critical float
    	set critical threshold (default 25)
  -d	set debug mode
  -ec2
    	check ec2 (default true)
  -elasticache
    	check elasticache
  -elasticsearch
    	check elasticsearch
  -rds
    	check rds
  -redshift
    	check redshift
  -w float
    	set warning threshold (default 50)
  -warning float
    	set warning threshold (default 50)
```

## Icinga2 example config

You can find an example config for icinga2 [here](doc/icinga2.conf)


CHANGELOG.md
# v0.2 (2021-07-29)

- check reservations from yesterday to today

# v0.1 (2021-06-11)

- Initial release


aws.go
/*
Copyright 2021 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"log"
	"os"
	"strconv"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/costexplorer"
)

const (
	defaultRegion = "eu-central-1"
)

func getReservationCoverage(service string, start string, end string) (*float64, error) {
	sess, conf := getAwsSession()
	srv := costexplorer.New(sess, conf)

	rs, err := srv.GetReservationCoverage(&costexplorer.GetReservationCoverageInput{
		TimePeriod: &costexplorer.DateInterval{
			End:   aws.String(end),
			Start: aws.String(start),
		},
		Filter: &costexplorer.Expression{
			Dimensions: &costexplorer.DimensionValues{
				Key: aws.String("SERVICE"),
				Values: []*string{
					getServiceName(service),
				},
			},
		},
	})
	if err != nil {
		return nil, err
	}
	rc, err := strconv.ParseFloat(*rs.Total.CoverageHours.CoverageHoursPercentage, 64)
	if err != nil {
		return nil, err
	}
	return &rc, nil
}

func getServiceName(key string) *string {
	if key == "ec2" {
		return aws.String("Amazon Elastic Compute Cloud - Compute")
	} else if key == "rds" {
		return aws.String("Amazon Relational Database Service")
	} else if key == "elasticache" {
		return aws.String("Amazon ElastiCache")
	} else if key == "redshift" {
		return aws.String("Amazon Redshift")
	} else if key == "elasticsearch" {
		return aws.String("Amazon Elasticsearch Service")
	}
	return nil
}

func getAwsSession() (*session.Session, *aws.Config) {
	conf := &aws.Config{Region: aws.String(defaultRegion)}
	if region := os.Getenv("REGION"); len(region) > 0 {
		log.Println("getAwsSession: set region to ", region)
		conf.Region = aws.String(region)
	}

	sess := session.Must(session.NewSession())
	if role := os.Getenv("ASSUME_ROLE"); len(role) > 0 {
		log.Println("getAwsSession: assume role ", role)
		creds := stscreds.NewCredentials(sess, role)
		conf.Credentials = creds
	}
	return sess, conf
}


check.go
/*
Copyright 2021 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"time"

	"github.com/lscheidler/go-nagios"
)

// Check data structure
type check struct {
	Critical      float64
	Debug         bool
	nagios        *nagios.Nagios
	Warning       float64
	ec2           bool
	rds           bool
	elasticache   bool
	redshift      bool
	elasticsearch bool
}

func new() *check {
	c := check{
		Critical:      25.0,
		Debug:         true,
		nagios:        nagios.New(),
		Warning:       50.0,
		ec2:           true,
		rds:           false,
		elasticache:   false,
		redshift:      false,
		elasticsearch: false,
	}
	parseArguments(&c)
	return &c
}

func (c *check) Run() {
	t := time.Now()
	today := t.Format("2006-01-02")
	yesterday := t.AddDate(0, 0, -1).Format("2006-01-02")

	if c.ec2 {
		c.checkService("ec2", "EC2NotCovered", yesterday, today)
	}
	if c.rds {
		c.checkService("rds", "RDSNotCovered", yesterday, today)
	}
	if c.elasticache {
		c.checkService("elasticache", "ElasticacheNotCovered", yesterday, today)
	}
	if c.redshift {
		c.checkService("redshift", "RedshiftNotCovered", yesterday, today)
	}
	if c.elasticsearch {
		c.checkService("elasticsearch", "ElasticsearchNotCovered", yesterday, today)
	}
}

func (c *check) checkService(key string, name string, start string, stop string) {
	if rc, err := getReservationCoverage(key, start, stop); err != nil {
		c.nagios.Unknown(err.Error())
	} else {
		c.nagios.CheckThreshold(name, 100-*rc, 100-c.Warning, 100-c.Critical)
	}
}

// Exit defer function
func (c *check) Exit() {
	c.nagios.Exit()
}


cli.go
/*
Copyright 2021 Lars Eric Scheidler

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	"os"
)

const (
	criticalUsage           = "set critical threshold"
	debugUsage              = "set debug mode"
	debugDefaultVal         = false
	warningUsage            = "set warning threshold"
	ec2Usage                = "check ec2"
	ec2DefaultVal           = true
	rdsUsage                = "check rds"
	rdsDefaultVal           = false
	elasticacheUsage        = "check elasticache"
	elasticacheDefaultVal   = false
	redshiftUsage           = "check redshift"
	redshiftDefaultVal      = false
	elasticsearchUsage      = "check elasticsearch"
	elasticsearchDefaultVal = false
)

func parseArguments(c *check) {
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "Usage of %s (%s):\n", os.Args[0], version)
		flag.PrintDefaults()

		examples := `
    `
		fmt.Fprintf(flag.CommandLine.Output(), "%s\n", examples)
	}

	flag.Float64Var(&c.Critical, "critical", 25.0, criticalUsage)
	flag.Float64Var(&c.Critical, "c", 25.0, criticalUsage)
	flag.BoolVar(&c.Debug, "d", debugDefaultVal, debugUsage)
	flag.Float64Var(&c.Warning, "warning", 50.0, warningUsage)
	flag.Float64Var(&c.Warning, "w", 50.0, warningUsage)
	flag.BoolVar(&c.ec2, "ec2", ec2DefaultVal, ec2Usage)
	flag.BoolVar(&c.rds, "rds", rdsDefaultVal, rdsUsage)
	flag.BoolVar(&c.elasticache, "elasticache", elasticacheDefaultVal, elasticacheUsage)
	flag.BoolVar(&c.redshift, "redshift", redshiftDefaultVal, redshiftUsage)
	flag.BoolVar(&c.elasticsearch, "elasticsearch", elasticsearchDefaultVal, elasticsearchUsage)

	flag.Parse()
}
